src/events.rs

```diff
@@ -0,0 +1,319 @@
+        }
+
+        if !path.exists() {
+            let parent = path.parent().expect("File needs a parent directory");
+
+            self.add_directory(&parent)
+                .map_err(|_e| Error::AddFailure)?;
+            self.pending_watched_files.insert(path.clone());
+        } else {
+            notify::Watcher::watch(&mut self.inner, &path, notify::RecursiveMode::NonRecursive)
+                .map_err(|_e| Error::AddFailure)?;
+
+            self.watched_files.insert(path.clone());
+        }
+
+        Ok(path)
+    }
+
+    fn handle_event(&mut self, event: &mut notify::Event) {
+        let mut retain_vec = vec![];
+
+        for path in &event.paths {
+            if self.pending_watched_files.contains(path) {
+                // TODO: should only do on events that imply file exists
+                let parent = path.parent().expect("Pending watched file needs a parent");
+                self.remove_directory(parent).unwrap();
+                self.pending_watched_files.remove(path);
+                let _res = self.add_file(path);
+            }
+
+            retain_vec.push(self.watched_files.contains(path));
+        }
+
+        let mut i = 0;
+        let paths = &mut event.paths;
+        paths.retain(|_| {
+            let retained = retain_vec[i];
+            i += 1;
+            retained
+        });
+    }
+
+    async fn next_event(&mut self) -> Option<Result<notify::Event, Error>> {
+        self.event_stream.next().await.map(|res| {
+            res.map_err(|e| {
+                match e.kind {
+                    notify::ErrorKind::Io(io_err) => io_err.into(),
+                    _ => {
+                        // Runtime event errors should only be std::io, but
+                        // need to handle this case anyway.
+                        io::Error::new(io::ErrorKind::Other, format!("Event error: {:?}", e)).into()
+                    }
+                }
+            })
+            .map(|mut event| {
+                self.handle_event(&mut event);
+                event
+            })
+        })
+    }
+}
+
+impl Default for MuxedEvents {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl FuturesStream for MuxedEvents {
+    type Item = notify::Event;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut task::Context<'_>,
+    ) -> task::Poll<Option<Self::Item>> {
+        use core::future::Future;
+        use futures_util::future::FutureExt;
+
+        let fut = self
+            .next_event()
+            .map(|val_opt| val_opt.map(|val_res| val_res.ok()).flatten());
+        pin_mut!(fut);
+        fut.poll(cx)
+    }
+}
+
+// TODO: maybe replace with crate `path-absolutize`
+fn absolutify(path: impl Into<PathBuf>, is_file: bool) -> PathBuf {
+    let path = path.into();
+
+    let (dir, maybe_filename) = if is_file {
+        (
+            path.parent().unwrap().to_path_buf(),
+            Some(path.file_name().unwrap().to_os_string()),
```

Comment:
Looking at `add_file` there isn't a check before this point to make sure that `path` isn't a directory, so it'd be very possible to hit this `unwrap`. I think this should either be an error, or `add_file` should have a "Panics" documentation section added specifying that this will panic if given a directory.

Reply:
Your intuition is spot-on; passing just `foo.txt` to the the `events` example exposes a bug where a parent can't be resolved here because `absolutify` failed to determine the absolute path without a `./`.

Looks like it can be fixed by replacing a None/empty parent with `env::current_dir`, but I'm leaning toward just using crate `path-absolutize` which also handles the dot[-less] logic correctly, and probably some more corner-cases I'm not thinking of. 

Will also return an error here for any other issues.